### ISSUE TITLE
No way to generate or parse timezone as produced by datetime.isoformat()

### ISSUE THREAD

**Issue by gvanrossum** [python/cpython#69142]
BPO | [24954](https://bugs.python.org/issue24954)
--- | :---
Nosy | @tim-one, @warsaw, @jcea, @abalkin, @vstinner, @jwilk, @mcepl, @merwok, @bitdancer, @karlcow, @flying-sheep, @mihaic, @Fak3, @berkerpeksag, @vadmium, @ztane, @boxed, @jstasiak, @offby1, @deronnax, @pbryan, @pganssle, @sirex, @mariocj89, @sonots
Dependencies | <li>bpo-31800: datetime.strptime: Support for parsing offsets with a colon</li>
Files | <li>[datetime.patch](https://bugs.python.org/file40675/datetime.patch "Uploaded as text/plain at 2015-10-04.19:58:23 by shanmbic")</li><li>[datetime2.patch](https://bugs.python.org/file40692/datetime2.patch "Uploaded as text/plain at 2015-10-05.22:44:10 by shanmbic")</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = 'https://github.com/abalkin'
closed_at = None
created_at = <Date 2015-08-28.22:34:41.660>
labels = ['extension-modules', 'easy', 'type-feature', '3.7']
title = 'No way to generate or parse timezone as produced by datetime.isoformat()'
updated_at = <Date 2018-07-05.19:25:05.940>
user = 'https://github.com/gvanrossum'
```

bugs.python.org fields:
```python
activity = <Date 2018-07-05.19:25:05.940>
actor = 'gvanrossum'
assignee = 'belopolsky'
closed = False
closed_date = None
closer = None
components = ['Extension Modules']
creation = <Date 2015-08-28.22:34:41.660>
creator = 'gvanrossum'
dependencies = ['31800']
files = ['40675', '40692']
hgrepos = ['319']
issue_num = 24954
keywords = ['patch', 'easy']
message_count = 24.0
messages = ['249305', '249815', '252284', '252294', '252364', '252366', '252449', '270530', '270531', '270543', '270573', '270575', '270579', '270826', '299213', '299237', '299239', '299241', '299333', '299339', '304596', '305015', '305016', '321108']
nosy_count = 36.0
nosy_names = ['tim.peters', 'barry', 'jcea', 'roysmith', 'belopolsky', 'nagle', 'vstinner', 'jwilk', 'mcepl', 'eric.araujo', 'Arfrever', 'r.david.murray', 'davydov', 'cvrebert', 'karlcow', 'SilentGhost', 'perey', 'flying sheep', 'mihaic', 'aymeric.augustin', 'Roman.Evstifeev', 'berker.peksag', 'martin.panter', 'piotr.dobrogost', 'kirpit', 'ztane', 'Anders.Hovm\xc3\xb6ller', 'jstasiak', 'Eric.Hanchrow', 'deronnax', 'pbryan', 'shanmbic', 'p-ganssle', 'sirex', 'mariocj89', 'sonots']
pr_nums = []
priority = 'normal'
resolution = None
stage = 'commit review'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue24954'
versions = ['Python 3.7']
```

</p></details>

**Comment by 502ba658-2675-4dda-a131-7932c4f0c00b** [comment]
Any progress here? I want `%:z` strftime.

By the way, ruby strptime works as:

```
irb(main):008:0* Time.strptime('2015-01-01T01:01:01+09:00', '%Y-%m-%dT%H:%M:%S%z')
=> 2015-01-01 01:01:01 +0900
irb(main):009:0> Time.strptime('2015-01-01T01:01:01+0900', '%Y-%m-%dT%H:%M:%S%z')
=> 2015-01-01 01:01:01 +0900
irb(main):010:0> Time.strptime('2015-01-01T01:01:01UTC', '%Y-%m-%dT%H:%M:%S%z')
=> 2015-01-01 01:01:01 UTC
irb(main):011:0> Time.strptime('2015-01-01T01:01:01Z', '%Y-%m-%dT%H:%M:%S%z')
=> 2015-01-01 01:01:01 UTC
irb(main):020:0> Time.strptime('2015-01-01T01:01:01Asia/Tokyo', '%Y-%m-%dT%H:%M:%S%z')
=> 2015-01-01 01:01:01 +0900
```

```
irb(main):015:0* Time.strptime('2015-01-01T01:01:01+09:00', '%Y-%m-%dT%H:%M:%S%Z')
=> 2015-01-01 01:01:01 +0900
irb(main):016:0> Time.strptime('2015-01-01T01:01:01+0900', '%Y-%m-%dT%H:%M:%S%Z')
=> 2015-01-01 01:01:01 +0900
irb(main):017:0> Time.strptime('2015-01-01T01:01:01UTC', '%Y-%m-%dT%H:%M:%S%Z')
=> 2015-01-01 01:01:01 UTC
irb(main):018:0> Time.strptime('2015-01-01T01:01:01Z', '%Y-%m-%dT%H:%M:%S%Z')
=> 2015-01-01 01:01:01 UTC
irb(main):019:0> Time.strptime('2015-01-01T01:01:01Asia/Tokyo', '%Y-%m-%dT%H:%M:%S%Z')
=> 2015-01-01 01:01:01 +0900
```

**Comment by gvanrossum** [comment]
Alexander, can you summarize the status of this issue? Maybe we can move forward for 3.7?

**Comment by abalkin** [comment]
This issue is waiting for the final decision on bpo-5288.  If sub-minute offsets support is accepted, I still don't think we need %::z because %:z can simply add non-zero :seconds as needed.  There are also some review comments on the latest patch that have not been addressed.  Overall this is an easy issue and it would be quite doable to get it resolved for 3.7.

**Comment by gvanrossum** [comment]
Please do! I have no opinion on %::z but maybe you can find inspiration in the Zen of Python. :-)

**Comment by 11ecd6cf-9a95-40f1-8f6c-6275193baf6b** [comment]
As the original author of the predecessor bug report (bpo-15873) in 2012, I would suggest that there's too much bikeshedding here. I filed this bug because there was no usable ISO8601 date parser available.  PyPi contained four slightly different buggy ones, and three more versions were found later.  

I suggested following RFC3339, "Date and Time on the Internet: Timestamps", section 5.6, which specifies a clear subset of ISO8601.  Five years later, I suggest just going with that. Fancier variations belong in non-standard libraries.

Date parsing should not be platform-dependent.  Using an available C library was convenient, but not portable. 

Let's get this done.

**Comment by abalkin** [comment]
John,

An RFC3339 parser is beyond the scope of this issue which is limited to adding str[fp]time code(s) to produce and consume RFC3339-formatted timezones.

We can still have fromisoformat() constructor implemented in 3.7, but someone needs to address the issues raised at bpo-15873.

**Comment by 8b1b57b4-b688-4865-84ed-b88aaa2990d4** [comment]
Wrote https://bugs.python.org/issue31800 without realising this issue was open (Thanks for bringing it up Martin Panter).

bpo-31800 basically just adds the ability to parse NN:NN to the already existing python isoformat function when %z is specified.
Rather than adding "%:z".

As available as well in linux strptime: http://man7.org/linux/man-pages/man3/strptime.3.html

I'd really like to see a way to parse isoformatted dates and this is the only thing in the middle. Happy to continue with the other issue/PR or help out here if needed.

Thanks! ^^

**Comment by abalkin** [comment]
I am going to merge [PR 4015](https://github.com/python/cpython/pull/4015) submitted for bpo-31800.  That issue asks to make %z strptime format accept offsets with : separators.  Given that a similar feature has been added to glibc several years ago, I don't see much need for bikeshedding.  This issue will remain open and will focus on strftime.

**Comment by gvanrossum** [comment]
SGTM.

**Comment by ThomasWaldmann** [comment]
Seems like there is still no way to have a custom format (like something including the suggested `%:z`) that generates the tzoffset in the same way as `datetime.isoformat()` would generate it: `+HH:MM`.

Thus, see topic of this issue, this can't be closed yet.

**Comment by ThomasWaldmann** [comment]
```
dt = datetime.now().astimezone()
dt.isoformat()
'2022-08-13T20:13:25.892454+02:00'
dt.strftime("%z")
'+0200'
dt.strftime("%:z")
':z'
```

**Comment by gvanrossum** [comment]
Indeed. There are some old patches attached to this diff. Would you like to clean one of those up and submit it as a PR?

**Comment by ThomasWaldmann** [comment]
@gvanrossum thanks for pointing me to the patches.

I had a look at them, but then decide to implement in a slightly different and easier way. PR coming soon.

**Comment by gvanrossum** [comment]
Closed by gh-95983. Thanks again TW!
